src/zasm.zig

```diff
@@ -0,0 +1,716 @@
+        flags.z_reserved16 = true;
+        flags.z_reserved25_31 = std.math.maxInt(u7);
+        flags.z_reserved32_63 = std.math.maxInt(u32);
+        break :blk @bitCast(u64, flags);
+    };
+
+    const ALL_NOT_RESERVED: u64 = ~ALL_RESERVED;
+
+    pub fn fromU64(value: u64) Cr4 {
+        return @bitCast(Cr4, value & ALL_NOT_RESERVED);
+    }
+
+    pub fn toU64(self: Cr4) u64 {
+        return @bitCast(u64, self) & ALL_NOT_RESERVED;
+    }
+
+    test {
+        try std.testing.expectEqual(@as(usize, 64), @bitSizeOf(Cr4));
+        try std.testing.expectEqual(@as(usize, 8), @sizeOf(Cr4));
+    }
+
+    comptime {
+        std.testing.refAllDecls(@This());
+    }
+};
+
+// ~~~ [ zig-x86_64/src/registers/model_specific.zig ] ~~~
+
+/// The Extended Feature Enable Register.
+pub const Efer = packed struct {
+    /// Enables the `syscall` and `sysret` instructions.
+    system_call_extensions: bool,
+
+    z_reserved1_7: u7,
+
+    /// Activates long mode, requires activating paging.
+    long_mode_enable: bool,
+
+    z_reserved9: bool,
+
+    /// Indicates that long mode is active.
+    long_mode_active: bool,
+
+    /// Enables the no-execute page-protection feature.
+    no_execute_enable: bool,
+
+    /// Enables SVM extensions.
+    secure_virtual_machine_enable: bool,
+
+    /// Enable certain limit checks in 64-bit mode.
+    long_mode_segment_limit: bool,
+
+    /// Enable the `fxsave` and `fxrstor` instructions to execute faster in 64-bit mode.
+    fast_fxsave_fxrstor: bool,
+
+    /// Changes how the `invlpg` instruction operates on TLB entries of upper-level entries.
+    translation_cache_extension: bool,
+
+    z_reserved16_31: u16,
+    z_reserved32_63: u32,
+
+    /// Read the current EFER flags.
+    pub fn read() Efer {
+        return Efer.fromU64(REGISTER.read());
+    }
+
+    /// Write the EFER flags, preserving reserved values.
+    ///
+    /// Preserves the value of reserved fields.
+    pub fn write(self: Efer) void {
+        REGISTER.write(self.toU64() | (REGISTER.read() & ALL_RESERVED));
+    }
+
+    const REGISTER = Msr(0xC000_0080);
+
+    const ALL_RESERVED: u64 = blk: {
+        var flags = std.mem.zeroes(Efer);
+        flags.z_reserved1_7 = std.math.maxInt(u7);
+        flags.z_reserved9 = true;
+        flags.z_reserved16_31 = std.math.maxInt(u16);
+        flags.z_reserved32_63 = std.math.maxInt(u32);
+        break :blk @bitCast(u64, flags);
+    };
+
+    const ALL_NOT_RESERVED: u64 = ~ALL_RESERVED;
+
+    pub fn fromU64(value: u64) Efer {
+        return @bitCast(Efer, value & ALL_NOT_RESERVED);
+    }
+
+    pub fn toU64(self: Efer) u64 {
+        return @bitCast(u64, self) & ALL_NOT_RESERVED;
+    }
+
+    test {
+        try std.testing.expectEqual(@as(usize, 64), @bitSizeOf(Efer));
+        try std.testing.expectEqual(@as(usize, 8), @sizeOf(Efer));
+    }
+
+    comptime {
+        std.testing.refAllDecls(@This());
+    }
+};
+
+fn Msr(comptime register: u32) type {
```

Comment:
The `Msr` function is private, since it does not have the `pub` attribute.

As such, all invocations to `Msr` are done from within this source file. However, their return value may be exported if stored in e.g.  a field of an exported struct.

This is the case with the `Efer` struct, which is exported (i.e. has `pub` attribute):

```zig
/// The Extended Feature Enable Register.
pub const Efer = packed struct {
    ...
    const REGISTER = Msr(0xC000_0080);
```